router.go

```diff
@@ -334,21 +333,45 @@ func (r *Router) Find(method, path string, c Context) {
 	cn := r.tree // Current node as root
 
 	var (
-		search  = path
-		child   *node         // Child node
-		n       int           // Param counter
-		nk      kind          // Next kind
-		nn      *node         // Next node
-		ns      string        // Next search
-		pvalues = ctx.pvalues // Use the internal slice so the interface can keep the illusion of a dynamic slice
+		search      = path
+		searchIndex = 0
+		n           int           // Param counter
+		pvalues     = ctx.pvalues // Use the internal slice so the interface can keep the illusion of a dynamic slice
 	)
 
-	// Search order static > param > any
-	for {
-		if search == "" {
-			break
+	// backtracking happens when we reach dead end when matching nodes in the router tree. To backtrack we set
+	// current node to parent to current node (this was previous node we checked) and choose next node kind to check.
+	// Next node kind relies on routing priority (static->param->any). So for example if there is no static node match we
+	// should check parent next sibling by kind (param). Backtracking itself does not check if there is next sibling this
+	// is left up to matching logic
+	backtrackToNextNodeKind := func(fromKind kind) (nextNodeKind kind, valid bool) {
+		previous := cn
+		cn = previous.parent
+		valid = cn != nil
+
+		// next node type by priority
+		// NOTE: by current implementation we never backtrack from any route so `previous.kind` value is here always static or any
+		// if this requirement is to change then for any route next kind would be static kind and this statement should be changed
+		nextNodeKind = previous.kind + 1
+
+		if fromKind == skind {
+			// when backtracking is done from static kind block we did not change search so nothing to restore
+			return
 		}
 
+		if previous.kind == skind {
+			searchIndex -= len(previous.prefix)
+			search = path[searchIndex:]
+		} else {
```

Comment:
Should either add a comment on what we process here, currently it's an any node, some code should fit for param node too.

Reply:
added comments for that part